EIPS/eip-5164.md

```diff
@@ -26,203 +26,286 @@ The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL
+
+`MultiChainMessageDispatcher`s MUST emit the `MessageBatchDispatched` event when a message batch is dispatched.
+
+`MultiChainMessageDispatcher`s MUST forward the message batch to the `MessageExecutor` on the `toChainId`.
+
+`MultiChainMessageDispatcher`s MUST use a unique `messageId` for each batch of messages.
+
+`MultiChainMessageDispatcher`s MUST return the `messageId` to allow the message sender to track the message batches.
+
+`MultiChainMessageDispatcher`s MAY require payment.
 
 ```solidity
-interface CrossChainRelayer {
-  error GasLimitTooHigh(
-    uint256 gasLimit,
-    uint256 maxGasLimit
-  );
+interface MultiChainMessageDispatcher is MessageDispatcher {
+  function dispatchMessageBatch(uint256 toChainId, Message[] calldata messages) external payable returns (bytes32 messageId);
 }
 ```
 
-### CrossChainExecutor
+```yaml
+- name: dispatchMessage
+  type: function
+  stateMutability: payable
+  inputs:
+    - name: toChainId
+      type: uint256
+    - name: messages
+      type: Message[]
+  outputs:
+    - name: messageId
+      type: bytes32
+```
+
+### MessageExecutor
 
-The `CrossChainExecutor` executes relayed calls. Developers must implement a `CrossChainExecutor` in order to execute messages on the receiving chain.
+The `MessageExecutor` executes dispatched messages and message batches. Developers must implement a `MessageExecutor` in order to execute messages on the receiving chain.
 
-The `CrossChainExecutor` will execute a nonce only once, but may execute nonces in any order. This specification makes no ordering guarantees, because messages may travel non-sequentially through the transport layer.
+The `MessageExecutor` will execute a messageId only once, but may execute messageIds in any order. This specification makes no ordering guarantees, because messages and message batches may travel non-sequentially through the transport layer.
 
 #### Execution
 
-`CrossChainExecutor`s SHOULD authenticate that the call has been performed by the bridge transport layer.
+`MessageExecutor`s SHOULD verify all message data with the bridge transport layer.
 
-`CrossChainExecutor`s MUST NOT execute a batch of calls more than once.
+`MessageExecutor`s MUST NOT successfully execute a message more than once.
+
+`MessageExecutor`s MUST revert the transaction when a message fails to be executed allowing the message to be retried at a later time.
 
 **Calldata**
 
-`CrossChainExecutor`s MUST append the ABI-packed (`nonce`, `sender`) to the calldata for each call being executed. It allows the receiver of the call to check the true `sender` of the transaction and use the `nonce` to apply any transaction ordering logic.
+`MessageExecutor`s MUST append the ABI-packed (`fromChainId`, `from`) to the calldata for each message being executed. This allows the receiver of the message to verify the cross-chain sender and the chain that the message is coming from.
```

Comment:
Should we also pass the `messageId` to allow the receiving contract to handle/verify transaction ordering?
That's why we were packing the `nonce` in the first place.

Reply:
Yeah, sounds good. It will be useful for pairing the execution event with external events.